src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexCoordinator.kt

```diff
@@ -261,10 +289,56 @@ class ManagedIndexCoordinator(
             if (it.value.shouldDeleteManagedIndexMetaData()) indicesToRemoveManagedIndexMetaDataFrom.add(it.value.index)
         }
 
-        updateManagedIndices(updateManagedIndicesRequests + indicesDeletedRequests, hasCreateRequests)
+        // check if newly created indices matching any ISM templates
+        val updateMatchingIndexReqs = getMatchingIndicesUpdateReqs(event.state(), event.indicesCreated())
+        if (updateMatchingIndexReqs.isNotEmpty()) hasCreateRequests = true
+
+        updateManagedIndices(updateManagedIndicesRequests + updateMatchingIndexReqs + indicesDeletedRequests, hasCreateRequests)
         clearManagedIndexMetaData(indicesToRemoveManagedIndexMetaDataFrom)
     }
 
+    /**
+     * build requests to create jobs for indices matching ISM templates
+     */
+    suspend fun getMatchingIndicesUpdateReqs(clusterState: ClusterState, indexNames: List<String>): List<DocWriteRequest<*>> {
+        val indexMetadatas = clusterState.metadata.indices
+        val templates = getISMTemplates()
+
+        val indexToMatchedPolicy = indexNames.map { indexName ->
+            indexName to findMatchingISMTemplate(templates, indexMetadatas[indexName])
+        }.toMap()
+
+        val updateManagedIndexReqs = mutableListOf<DocWriteRequest<*>>()
+        indexToMatchedPolicy.filterNotNullValues()
+            .forEach { (index, policyID) ->
+            val indexUuid = indexMetadatas[index].indexUUID
+            if (indexUuid != null) {
+                logger.info("auto manage index $index to policy $policyID")
+                updateManagedIndexReqs.add(
+                    managedIndexConfigIndexRequest(index, indexUuid, policyID, jobInterval))
+            }
+        }
+
+        return updateManagedIndexReqs
+    }
+
+    suspend fun getISMTemplates(): Map<String, ISMTemplate> {
+        val searchRequest = SearchRequest()
+            .source(
+                SearchSourceBuilder().query(
+                    QueryBuilders.existsQuery(ISM_TEMPLATE_FIELD)))
+            .indices(INDEX_MANAGEMENT_INDEX)
+
+        return try {
+            val response: SearchResponse = client.suspendUntil { search(searchRequest, it) }
+            ismTemplatesFromSearchResponse(response).filterNotNullValues()
+        } catch (ex: IndexNotFoundException) {
+            emptyMap()
+        } catch (ex: ClusterBlockException) {
+            emptyMap()
```

Comment:
The reason I catch this here is because I see this exception when I run `./gradlew run` locally, the cluster is blocked for some time when it's being brought up.

For other situation, I am thinking if cluster is blocked, this whole ISM template function is also not functioning since there should not be any new indices created, so I just return a emptyMap()

will leave this conversation open for other reviewers to discuess

Reply:
wrapped getISMTemplates call within newly created indices event